Source/WebGPU/WGSL/TypeCheck.cpp

```diff
@@ -0,0 +1,321 @@
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
+ * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
+ * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "TypeCheck.h"
+
+#include "AST.h"
+#include "ASTVisitor.h"
+#include "ContextProviderInlines.h"
+#include <wtf/SetForScope.h>
+#include <wtf/Vector.h>
+
+namespace WGSL {
+
+class UnificationContext;
+
+struct Type {
+    enum Kind {
+        Variable,
+        Primitive,
+        Constructor,
+        Application,
+        Function,
+    };
+};
+
+using SubstituionCallback = std::function<void(Type*)>;
+
+class TypeChecker : public AST::Visitor, public ContextProvider<Type*> {
+    friend class UnificationContext;
+
+public:
+    TypeChecker(AST::ShaderModule& shaderModule)
+        : m_shaderModule(shaderModule)
+    {
+    }
+
+    void check();
+
+    void visit(AST::Structure&) override;
+    void visit(AST::Variable&) override;
+    void visit(AST::Function&) override;
+    void visit(AST::Statement&) override;
+    void visit(AST::AssignmentStatement&) override;
+    void visit(AST::ReturnStatement&) override;
+    void visit(AST::Expression&) override;
+    void visit(AST::FieldAccessExpression&) override;
+    void visit(AST::IndexAccessExpression&) override;
+    void visit(AST::BinaryExpression&) override;
+    void visit(AST::IdentifierExpression&) override;
+    void visit(AST::TypeName&) override;
+
+private:
+    Type* infer(AST::Expression&);
+    Type* variable();
+    Type* resolve(AST::TypeName&);
+    void inferred(Type*);
+    void unify(Type*, Type*);
+
+    void substitute(Type*, SubstituionCallback&&);
+
+    AST::ShaderModule& m_shaderModule;
+    Type* m_inferredType { nullptr };
+    UnificationContext* m_unificationContext { nullptr };
+
+    // FIXME: remove this once we start allocating types
+    Type m_type;
+};
+
+class UnificationContext {
+public:
+    UnificationContext(TypeChecker* typeChecker)
+        : m_setForScope(typeChecker->m_unificationContext, this)
+    {
+    }
+
+    ~UnificationContext()
+    {
+        solveConstraints();
+        runSubstitutions();
+    }
+
+    void unify(Type* lhs, Type* rhs)
+    {
+        m_constraints.append({ lhs, rhs });
+    }
+
+    void substitute(Type* type, SubstituionCallback&& callback)
+    {
+        m_substitutions.append({ type, WTFMove(callback) });
+    }
+
+private:
+    void solveConstraints()
+    {
+        // FIXME: implement this
```

Comment:
Is this really so complicated that it has to be described as a graph of constraints? When we were designing WGSL we tried pretty hard to make sure that every expression had a well-defined type that was context-independent. Why can't we just naively make this a single traversal from leaves to root, annotating types at each node?

Reply:
That's a good point, maybe it's unnecessary? I often find it a little easier to reason about it this way, and I was considering the combination of overload resolution with the conversion rules. I'll keep it this way for now, but I'll try to remove it once I get to implementing the actual checking.